api/opensource/mfx_dispatch/linux/mfxloader.cpp

```diff
@@ -0,0 +1,582 @@
+  mfx_res = ctx.front().Load(uid, version, path);
+  if (MFX_ERR_NONE != mfx_res) {
+    return mfx_res;
+  }
+
+  try {
+    std::lock_guard<std::mutex> lock(m_guard);
+    m_plugins.splice(m_plugins.end(), ctx);
+  } catch(...) {
+    return MFX_ERR_UNKNOWN;
+  }
+
+  return MFX_ERR_NONE;
+}
+
+mfxStatus LoaderCtx::UnloadPlugin(const mfxPluginUID& uid)
+{
+  mfxStatus mfx_res = MFX_ERR_NONE;
+
+  try {
+    std::list<PluginCtx> ctx;
+    {
+      std::lock_guard<std::mutex> lock(m_guard);
+      for (auto it = m_plugins.begin(); it != m_plugins.end(); ++it) {
+        if (it->getUID() == uid) {
+          mfx_res = it->Unload();
+          if (MFX_ERR_NONE != mfx_res) {
+            return mfx_res;
+          }
+
+          ctx.splice(ctx.end(), m_plugins, it);
+          break;
+        }
+      }
+    }
+  } catch(...) {
+    return MFX_ERR_UNKNOWN;
+  }
+
+  return MFX_ERR_NONE;
+}
+
+} // namespace MFX
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+mfxStatus MFXInit(mfxIMPL impl, mfxVersion *ver, mfxSession *session)
+{
+    mfxInitParam par;
+
+    memset(&par, 0, sizeof(par));
+    par.Implementation = impl;
+    if (ver) {
+        par.Version = *ver;
+    } else {
+        par.Version.Major = MFX_VERSION_MAJOR;
+        par.Version.Minor = MFX_VERSION_MINOR;
+    }
+
+    return MFXInitEx(par, session);
+}
+
+mfxStatus MFXInitEx(mfxInitParam par, mfxSession *session)
+{
+  if (!session) return MFX_ERR_NULL_PTR;
+
+  std::unique_ptr<MFX::LoaderCtx> loader;
+
+  try {
+    loader.reset(new MFX::LoaderCtx{});
+  } catch(...) {
+    return MFX_ERR_MEMORY_ALLOC;
+  }
+
+  mfxStatus mfx_res = loader->Init(par);
+  if (MFX_ERR_NONE == mfx_res) {
+    *session = (mfxSession)loader.release();
+  } else {
+    *session = nullptr;
+  }
+
+  return mfx_res;
+}
+
+mfxStatus MFXClose(mfxSession session)
+{
+  std::unique_ptr<MFX::LoaderCtx> loader((MFX::LoaderCtx*)session);
+
+  if (!loader.get()) return MFX_ERR_INVALID_HANDLE;
+
+  mfxStatus mfx_res = loader->Close();
+  if (mfx_res == MFX_ERR_UNDEFINED_BEHAVIOR) {
+    // It is possible, that there is an active child session.
+    // Can't unload library in this case.
+    loader.release();
+  }
+  return mfx_res;
+}
+
+mfxStatus MFXVideoUSER_Load(mfxSession session, const mfxPluginUID *uid, mfxU32 version)
+{
+  MFX::LoaderCtx* loader = (MFX::LoaderCtx*)session;
+  std::string path;
+
+  if (!loader) return MFX_ERR_INVALID_HANDLE;
+  if (!uid) return MFX_ERR_NULL_PTR;
+
+  {
+    std::lock_guard<std::mutex> lock(MFX::g_GlobalCtx.m_mutex);
+
+    auto find_uid = [](const mfxPluginUID& puid) {
+      return std::find_if(
+        MFX::g_GlobalCtx.m_plugins.begin(),
+        MFX::g_GlobalCtx.m_plugins.end(),
+        [&puid](MFX::PluginInfo& item){ return item.getUID() == puid; }
+        );
+    };
+
+    // checking whether we have already cached this plugin for this application
+    auto it = find_uid(*uid);
+    if (it == MFX::g_GlobalCtx.m_plugins.end() && !MFX::g_GlobalCtx.m_plugins.size()) {
```

Comment:
use g_GlobalCtx.m_plugins.empty() to check that container is (isn't) empty

Reply:
yep. thank you for the catch.